### PR TITLE
Encode hostname directly instead of netloc

### PR DESCRIPTION
**Pull request by Laerte**
This fix the issue 
- #171

### CODE REVIEW

File: w3lib/url.py [hunk 1]
```diff
@@ -95,10 +95,26 @@ def safe_url_string(
...
     # or missing labels (e.g. http://.example.com)
     try:
-        netloc_bytes = parts.netloc.encode("idna")
+        # When we have hostname we use it instead of netloc directly
+        if hostname:
+            if isinstance(username, str) and isinstance(password, str):
+                netloc_bytes += f'{username}:{password}@'.encode("idna")
+            netloc_bytes += hostname.encode("idna")
+            if port_number:
+                netloc_bytes += f":{port_number}".encode("idna")
```
Comment (Gallaecio) [inline w3lib/url.py:115]: You are encoding everything as `idna`, but the [StackOverflow](https://stackoverflow.com/questions/804336/best-way-to-convert-a-unicode-url-to-ascii-utf-8-percent-escaped-in-python/804380#804380) answer that @wRAR linked in the issue being addressed only does so for the `hostname`. Are you sure those other parts also need to be encoded as `idna`?

File: tests/test_url.py [hunk 2]
```diff
@@ -1104,6 +1104,18 @@ def test_scheme_case_insensitive(self):
...
+        self.assertEqual(safe_url_string('http://新华网.中国:80'),
+                         'http://xn--xkrr14bows.xn--fiqs8s:80')
+
+    def test_safe_url_string_encode_idna_domain_with_username_password_and_port_number(self):
+        self.assertEqual(safe_url_string('ftp://admin:admin@新华网.中国:21'),
+                         'ftp://admin:admin@xn--xkrr14bows.xn--fiqs8s:21')
+
+    def test_safe_url_string_encode_idna_domain_with_username_without_password_and_port_number(self):
+        self.assertEqual(safe_url_string('ftp://admin:@新华网.中国:21'),
+                         'ftp://admin:@xn--xkrr14bows.xn--fiqs8s:21')
```
Comment (Gallaecio) [inline tests/test_url.py:1117]: I believe this tests an empty password. No password would mean no colon: `…//admin@…`

File: w3lib/url.py (hunk 1 above)
Comment (Laerte) [reply w3lib/url.py]: Changed, thanks.

File: tests/test_url.py (hunk 2 above)
Comment (Laerte) [reply tests/test_url.py]: Did know about that, fixed and rename the test.

File: tests/test_url.py [hunk 3]
```diff
@@ -1104,6 +1104,35 @@ def test_scheme_case_insensitive(self):
...
+            "http://jap%C3%A3o:n%C3%A3o@xn--i-7iq.ws:8000/",
+        )
+
+    def test_safe_url_string_encode_idna_domain_with_username_and_empty_password_and_port_number(
+        self,
+    ):
+        self.assertEqual(
+            safe_url_string("ftp://admin:@新华网.中国:21"),
+            "ftp://admin@xn--xkrr14bows.xn--fiqs8s:21",
+        )
```
Comment (Gallaecio) [inline tests/test_url.py:1134]: Can you also add a test for an empty password?

I suspect it will fail, because in that case I think `part.password` is `None`, but your current code seems to ignore the user if `part.password` is not a string.

Comment (Laerte) [reply tests/test_url.py:1134]: This is already the test for empty password, i'm missing something?

File: w3lib/url.py [hunk 4]
```diff
@@ -95,10 +95,29 @@ def safe_url_string(
...
+    netloc_bytes = b""
+
     # IDNA encoding can fail for too long labels (>63 characters)
     # or missing labels (e.g. http://.example.com)
     try:
-        netloc_bytes = parts.netloc.encode("idna")
+        # When we have hostname we use it instead of netloc directly
+        if hostname:
+            if isinstance(username, str) and isinstance(password, str):
+                netloc_bytes += quote(username, _safe_chars).encode(encoding)
```
Comment (Gallaecio) [inline w3lib/url.py:112]: Do we need `quote` here? I’m mostly concerned because `_safe_chars` contains some characters that may not be safe here, such as `:` for a username.

File: tests/test_url.py (hunk 3 above)
Comment (Laerte) [reply tests/test_url.py:1134]: Done.

File: w3lib/url.py (hunk 4 above)
Comment (Laerte) [reply w3lib/url.py:112]: I have the same question, i follow the SO answer, what do you think? Or maybe we could remove the `_safe_chars` param.

Comment (Laerte) [reply w3lib/url.py:112]: @Gallaecio Do you think worth add test for empty username or this is a edgy case?

**Comment by Gallaecio** [comment]
I need to delay my review until I have the time to carefully look into the potential issues with encoding and decoding each separate field.

**Comment by Gallaecio** [comment]
@Laerte Sorry for the wait, I finally had a proper look.

I have created a pull request against your branch, https://github.com/Laerte/w3lib/pull/1, please have a look when you get a chance.

`quote` was indeed needed for username and password, but `_safe_chars` was not the right character allow-list for those fields.

**Comment by Laerte** [comment]
@Gallaecio Thank you! Do you think we can merge this to `master` or need more changes?

**Review by Gallaecio — APPROVED** [review]
It looks good to me. I’ll try and get someone else to have a look as well before we merge.

**Comment by Laerte** [comment]
@Gallaecio Any update on this one?

**Review by wRAR — APPROVED** [review]
